okapi-core/src/test/java/org/folio/okapi/util/DepResolutionTest.java

```diff
@@ -220,23 +235,247 @@ public void test4(TestContext context) {
+      context.assertEquals("enable moduleD-1.1.0 failed: interface int required by module moduleD-1.1.0 not found", res.cause().getMessage());
+      async.complete();
+    });
+  }
+
+  // install optional with existing interface that needs upgrading
+  @Test
+  public void testInstallOptionalExistingModule(TestContext context) {
+    Async async = context.async();
+
+    Map<String, ModuleDescriptor> modsAvailable = new HashMap<>();
+    modsAvailable.put(mdA100.getId(), mdA100);
+    modsAvailable.put(mdA110.getId(), mdA110);
+    modsAvailable.put(mdD100.getId(), mdD100);
+    modsAvailable.put(mdD110.getId(), mdD110);
+    modsAvailable.put(mdE100.getId(), mdE100);
+
+    Map<String, ModuleDescriptor> modsEnabled = new HashMap<>();
+    modsEnabled.put(mdA100.getId(), mdA100);
+
+    List<TenantModuleDescriptor> tml = new LinkedList<>();
+    TenantModuleDescriptor tm = new TenantModuleDescriptor();
+    tm.setAction(TenantModuleDescriptor.Action.enable);
+    tm.setId(mdD110.getId());
+    tml.add(tm);
+
+    DepResolution.installSimulate(modsAvailable, modsEnabled, tml, res -> {
+      context.assertTrue(res.succeeded());
+      logger.debug("tml result = " + Json.encodePrettily(tml));
+      context.assertEquals(2, tml.size());
+      context.assertEquals("moduleA-1.1.0", tml.get(0).getId());
+      context.assertEquals("moduleA-1.0.0", tml.get(0).getFrom());
+      context.assertEquals("enable", tml.get(0).getAction().name());
+      context.assertEquals("moduleD-1.1.0", tml.get(1).getId());
+      context.assertEquals(null, tml.get(1).getFrom());
+      context.assertEquals("enable", tml.get(1).getAction().name());
+      async.complete();
+    });
+  }
+
+  // upgrade base dependency which is still compatible with optional interface
+  @Test
+  public void testInstallOptionalExistingModule2(TestContext context) {
+    Async async = context.async();
+
+    Map<String, ModuleDescriptor> modsAvailable = new HashMap<>();
+    modsAvailable.put(mdA100.getId(), mdA100);
+    modsAvailable.put(mdA110.getId(), mdA110);
+    modsAvailable.put(mdD100.getId(), mdD100);
+    modsAvailable.put(mdD110.getId(), mdD110);
+
+    Map<String, ModuleDescriptor> modsEnabled = new HashMap<>();
+    modsEnabled.put(mdA100.getId(), mdA100);
+    modsEnabled.put(mdD100.getId(), mdD100);
+
+    List<TenantModuleDescriptor> tml = new LinkedList<>();
+    TenantModuleDescriptor tm = new TenantModuleDescriptor();
+    tm.setAction(TenantModuleDescriptor.Action.enable);
+    tm.setId(mdA110.getId());
+    tml.add(tm);
+
+    DepResolution.installSimulate(modsAvailable, modsEnabled, tml, res -> {
+      context.assertTrue(res.succeeded());
+      logger.debug("tml result = " + Json.encodePrettily(tml));
+      context.assertEquals(1, tml.size());
+      context.assertEquals("moduleA-1.1.0", tml.get(0).getId());
+      context.assertEquals("moduleA-1.0.0", tml.get(0).getFrom());
+      context.assertEquals("enable", tml.get(0).getAction().name());
```

Comment:
Yes, no moduleD upgrade is needed because moduleA's new interface is compatible (int-1.0 -> int-1.1).

Can you add another test where moduleA-1.0.0 is upgraded to moduleA-2.0.0 with int-2.0 and this triggers the moduleD upgrade from moduleD-1.0.0 to moduleD-2.0.0 because moduleD-2.0.0 has optional dependency on int-2.0?

Reply:
Good one. That was a reverse dependency not catched.. Now fixed.